clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -500,11 +521,21 @@ public Record next() {
                 throw new NoSuchElementException();
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // Validate that the actual size of the batch is equal to declared size
+                // by checking that after reading declared number of items, there no items left
+                // (overflow case, i.e. reading past buffer end is checked elsewhere).
+                if (!hasRemaining())
+                    throw new InvalidRecordException("Incorrect declared batch size, records still remaining in file");
+            }
+            return rec;
         }
 
         protected abstract Record readNext(long baseOffset, long baseTimestamp, int baseSequence, Long logAppendTime);
 
+        protected abstract boolean hasRemaining();
```

Comment:
Thinking about this name a little more... Since the InputStream version uses `read()`, this method cannot be used as generally as its name suggests. We should try to convey the fact that this check should only be used when the presumed end has been reached. Maybe something like `ensureNoneRemaining`?